### PR TITLE
feature: Add support for seq2seq

### PR DESCRIPTION
**Pull request by JulesBelveze**
This PR introduces several enhancements and refactoring changes, particularly focusing on sequence-to-sequence tasks and model training support.

## Features:
- Added seq2seq distillation support to leverage knowledge distillation for sequence-to-sequence models (1eebdcb).
- Support for T5 models training has been implemented, extending the library's versatility for different model types (7957fb3).
- Added language model and summarization scorers to provide a range of evaluation metrics (b6f590f).
- Introduced the Seq2SeqTransformerDataModule to handle data related operations for sequence-to-sequence tasks (3f82a99).

## Refactoring:
- The code has been restructured to improve readability and formatting, making it more maintainable and easier to understand (a41d871).
- Updated the perplexity calculation and logging to reflect more accurate measurements (7862746).
- Refactored the distiller classes to accommodate the sequence classification task more effectively (7da19ed).
- Improved base classes for consistency and enhancement in handling sequence classification and seq2seq tasks (3e4c7b0, 56bc51e).
- Restructured scorers for better organization and usage within the library (75a903d).
- Transitioned to using the BaseSequenceClassificationTransformerModule to streamline the underlying model handling (d54fa12).
- Separated base classes for Transformer modules to establish a clearer hierarchy and improve the extension capabilities of the library (94c9b4d).

## Other Changes:
- Added evaluate as a dependency to assist in the model evaluation process (bc7d95e).
- Enhanced the BaseDataModule with detailed docstrings, improving the documentation and understandability of the data handling process (6ad0d0d).

  These updates aim to solidify `bert-squeeze`'s position as a versatile tool for training and deploying sequence-to-sequence models while ensuring high code quality and performance

### CODE REVIEW

File: bert_squeeze/data/modules/transformer_module.py [hunk 1]
```diff
@@ -180,3 +178,132 @@ def featurize(self) -> datasets.DatasetDict:
...
+            collate_fn=self._collate_fn(),
+            batch_size=self.eval_batch_size,
+            drop_last=True,
+        )
+
+    def _collate_fn(self):
+        """Helper function to merge a list of samples into a batch of Tensors"""
+
+        def _collate(examples):
+            return self.tokenizer.pad(examples, return_tensors="pt")
```
Comment (tarekziade) [inline bert_squeeze/data/modules/transformer_module.py:307]: what does this do? I thought the padding was done directly when you tokenize the dataset earluer?

File: bert_squeeze/utils/scorers/lm_scorer.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        else:
+            self.tokenizer = None
+        self.do_mismatch = do_mismatch
+
+        self.losses = defaultdict(list)
+        self.metrics = defaultdict(list)
+        self.mismatches = []
+
+    @property
+    def ppl(self):
```
Comment (tarekziade) [inline bert_squeeze/utils/scorers/lm_scorer.py:28]: is `ppl` stands for `perplexity` ? I would use the full name for clarity as the shorter version is not that shorter

File: bert_squeeze/utils/scorers/lm_scorer.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        loss: torch.Tensor = None,
+        predicted_tokens: torch.Tensor = None,
+        labels: torch.Tensor = None,
+        input_ids: torch.Tensor = None,
+    ):
+        """"""
+        with torch.no_grad():
+            self.losses["global"].append(loss.cpu().numpy())
+
+            ppl = torch.clip(torch.exp(loss), max=1e8)
```
Comment (tarekziade) [inline bert_squeeze/utils/scorers/lm_scorer.py:48]: maybe push `1e8` in a CONSTANT at the top of the file

File: tests/assistants/test_train_assistant.py [hunk 4]
```diff
@@ -147,3 +147,34 @@ def test_adapter_training(self):
             train_dataloaders=train_dataloader,
             val_dataloaders=test_dataloader,
         )
+
+
+class TestSeq2SeqTraining:
+    def test_t5_summarization(self):
```
Comment (tarekziade) [inline tests/assistants/test_train_assistant.py:153]: yay!

**Review by tarekziade — APPROVED** [review]

File: bert_squeeze/data/modules/transformer_module.py (hunk 1 above)
Comment (JulesBelveze) [reply bert_squeeze/data/modules/transformer_module.py:307]: This basically padded the within the batch. We don't really want to always pad to the max length as it will lead to big inputs.

Comment (JulesBelveze) [reply bert_squeeze/data/modules/transformer_module.py:307]: I'll set the tokenizer to never pad though, I'm seeing it's in "longest" mode now
